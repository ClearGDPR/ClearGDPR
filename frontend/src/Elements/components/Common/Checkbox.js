import React from 'react';
import PropTypes from 'prop-types';

import styles from 'theme/Checkbox.scss';

class Checkbox extends React.PureComponent {
  render() {
    const exStyles = this.props.styles || {};

    return (
      <label className={styles.labelWrapper} htmlFor={this.props.id} name={this.props.name}>
        <input
          className={styles.checkbox}
          type="checkbox"
          id={this.props.id}
          required={this.props.required}
          onChange={this.props.onChange}
        />
        <span style={exStyles.label} className={styles.label}>
          {this.props.label}
        </span>
      </label>
    );
  }
}

Checkbox.propTypes = {
  id: PropTypes.any,
  name: PropTypes.string,
  label: PropTypes.string,
  required: PropTypes.bool,
  styles: PropTypes.object,
  onChange: PropTypes.func
};

export default Checkbox;
