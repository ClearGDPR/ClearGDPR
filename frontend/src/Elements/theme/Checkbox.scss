@import 'variables';

.labelWrapper {
  display: flex;
  align-items: center;
}

.label {
  margin-left: 16px;

  &::-moz-selection {
    background: none;
  }

  &::selection {
    background: none;
  }
}

.checkbox,
.checkbox[type='checkbox'] {
  cursor: pointer;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  outline: 0;
  height: 16px;
  width: 16px;
  padding: 11px;
  background: white;
  position: relative;
  border: 2px solid #aaa;
  border-radius: 3px;

  &:checked {
    border-color: $green;
    background: $green;

    &::after {
      content: '\2714';
      font-size: 14px;
      color: white;
      position: absolute;
      top: 2px;
      left: 5px;
    }
  }
}
