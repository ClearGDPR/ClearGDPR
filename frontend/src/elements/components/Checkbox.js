import React from 'react';
import PropTypes from 'prop-types';
import styles from '../theme/Checkbox.scss';

class Checkbox extends React.PureComponent {
  render() {
    const exStyles = this.props.styles || {};

    return (
      <label
        className={styles.labelWrapper}
        htmlFor={this.props.id}
        name={this.props.name}
        onChange={this.props.onChange}
      >
        <input
          className={styles.checkbox}
          type="checkbox"
          id={this.props.id}
          required={this.props.required}
        />
        <span style={exStyles.label} className={styles.label}>
          {this.props.label}
        </span>
      </label>
    );
  }
}

Checkbox.propTypes = {
  styles: PropTypes.object,
  label: PropTypes.string
};

export default Checkbox;
