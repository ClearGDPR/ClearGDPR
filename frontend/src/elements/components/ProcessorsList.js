import React from 'react';
import PropTypes from 'prop-types';
import Processor from './Processor';

class ProcessorsList extends React.PureComponent {
  render() {
    const { processors } = this.props;

    return (
      <React.Fragment>
        {processors.length > 0
          ? processors.map(p => (
              <Processor
                key={p.id}
                processor={p}
                onProcessorChange={this.props.onProcessorChange}
              />
            ))
          : 'No processors'}
      </React.Fragment>
    );
  }
}

ProcessorsList.propTypes = {
  styles: PropTypes.object // TODO: not necessarily useful here
};

export default ProcessorsList;
