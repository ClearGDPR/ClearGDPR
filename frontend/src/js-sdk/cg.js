import 'es6-promises';
import Config from './config';
import Modules from './modules';
import { SDKError } from './common/exceptions';
import { mergeOptions } from './utils/helpers';

/**
 * Creates a new CG ClearGDPR API client
 * @constructor
 * @classdesc The CG SDK class provides methods to manage user's data
 * @param {Object}  options
 * @param {String}  options.apiKey Key used to sign up to the API
 * @param {String}  options.apiUrl URL to the API server without trailing slash
 * @param {String}  options.apiVersion Version of the API, like `v1`
 * @param {String}  options.apiTimeout Timeout used when a Request is made to the API
 * @param {Boolean} options.debug Flag to enable or disable complete error stack trace
 * @see {@link https://docs.cleargdpr.com/api}
 */
export class CG {
  constructor({ apiKey, apiUrl }) {
    // Enforce accessToken option to be sent.
    if (!apiKey) {
      throw new SDKError('API Key is required.');
    }

    if (!apiUrl) {
      throw new SDKError('API URL is required.');
    }

    this._config = mergeOptions(Config, { apiUrl });
    this.setAPIKey(apiKey);
    this.registerModules();
  }

  registerModules() {
    Object.keys(Modules).map(key => {
      const endpoint = new Modules[key](this);
      Object.assign(this, { [key]: endpoint });
    });
  }

  getAccessToken() {
    return this._config.auth;
  }

  setAccessToken(accessToken) {
    if (accessToken) {
      // TODO: Should validate quicker if its a valid JWT
      // TODO: How to verify JWT with our Server + Consumer
      this._setConfigField('auth', `Bearer ${accessToken}`);
    }
  }

  getAPIKey() {
    return this._config.auth;
  }

  setAPIKey(apiKey) {
    if (apiKey) {
      this._setConfigField('apiKey', apiKey);
    }
  }

  getBaseURL() {
    return this._config.apiVersion
      ? `${this._config.apiUrl}/${this._config.apiVersion}`
      : this._config.apiUrl;
  }

  _setConfigField(key, value) {
    this._config[key] = value;
  }
}
