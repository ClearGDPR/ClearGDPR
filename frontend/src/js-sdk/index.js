/*
 * cg-js-sdk
 * https://www.cleargdpr.com/
 * 
 */
export { CG } from './cg';
