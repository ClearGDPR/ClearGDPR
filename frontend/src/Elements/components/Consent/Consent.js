import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import _ from 'lodash';

import Checkbox from 'components/Common/Checkbox';
import PopoverView from 'components/Common/Views/Popover';
import ProcessorsList from 'components/Processors/ProcessorsList';
import Subject from 'contexts/Subject';

class Consent extends React.PureComponent {
  state = {
    processors: [],
    showProcessors: false
  };

  async componentDidMount() {
    await this.props.subject.fetchProcessors();
    const { processors, isGuest } = this.props.subject;
    this.setState({ processors, isGuest });

    this.form = ReactDOM.findDOMNode(this).parentNode;
    this.props.subject.onEntrance(this.handleGiveConsent);
  }

  componentWillUnmount() {
    this.props.subject.clearEntranceListeners();
  }

  toggleProcessors = e => {
    e.preventDefault();

    this.setState(prevState => ({
      showProcessors: !prevState.showProcessors
    }));
  };

  handleGiveConsent = async () => {
    let processors = this.state.processors.slice();

    // Get Data from parent form
    const data = Array.prototype.slice
      .call(this.form.querySelectorAll('[data-cleargdpr="true"]'), 0)
      .reduce((result, formField) => {
        if (formField.value) {
          result[formField.name] = formField.value;
        }
        return result;
      }, []);

    try {
      await this.giveConsent(data, processors.filter(p => p.enabled).map(p => p.id));
      if (this.props.options.onSuccessCallback) {
        this.props.options.onSuccessCallback();
      }
    } catch (e) {
      console.log(e);
    }
  };

  onProcessorChange = processor => {
    let processors = this.state.processors.slice();
    let index = _.findIndex(processors, { id: processor.id });
    if (index !== -1) {
      processors.splice(index, 1, processor);
      this.setState({ processors });
    }
  };

  giveConsent = async (data = {}, processors) => {
    this.props.subject.giveConsent({
      personalData: { ...data },
      processors: [...processors]
    });
  };

  render() {
    return (
      <div>
        <Checkbox
          styles={this.props.options.styles}
          label={this.props.options.label}
          onChange={this.onChange}
          required={this.props.options.required}
          name="cleargdpr-give-consent"
        />
        <div style={{ position: 'relative', padding: '10px 0 0 40px' }}>
          <button
            className="button is-primary is-outlined is-small"
            onClick={this.toggleProcessors}
          >
            <span role="img" aria-labelledby="lock">
              ðŸ”’
            </span>
            Config Processors
          </button>
          {this.state.showProcessors ? (
            <PopoverView open={this.state.showProcessors} onClose={this.toggleProcessors}>
              <ProcessorsList
                processors={this.state.processors}
                onProcessorChange={this.onProcessorChange}
              />
            </PopoverView>
          ) : null}
        </div>
      </div>
    );
  }
}

Consent.propTypes = {
  options: PropTypes.object,
  subject: PropTypes.instanceOf(Subject)
};

export default Consent;
