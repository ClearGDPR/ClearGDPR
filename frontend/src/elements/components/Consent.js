import React from 'react';
import ReactDOM from 'react-dom';
import ProcessorsList from './ProcessorsList';
import Checkbox from './Checkbox';
import PopoverView from './Views/Popover';
import _ from 'lodash';
import config from '../../config';

const { API_BASE } = config;

class Consent extends React.PureComponent {
  state = {
    registered: false,
    processors: [],
    showProcessors: false
  };

  componentDidMount() {
    // TODO: Improve cache, save to elements store, add localStorage
    if (!this.state.processors.lenght) {
      window.cg.Subject.getProcessors()
        .then(processors => {
          processors.map(p => {
            p.enabled = true;
            return p;
          });
          this.setState({ processors });
        })
        .catch(err => {
          console.log('failure', err);
        });
    }

    const form = ReactDOM.findDOMNode(this).parentNode;
    form.addEventListener('submit', this.handleGiveConsent.bind(this));
  }

  componentWillUnmount() {
    const form = ReactDOM.findDOMNode(this).parentNode;
    form.removeEventListener('submit', () => {});
  }

  toggleProcessors(e) {
    e.preventDefault();

    this.setState(prevState => ({
      showProcessors: !prevState.showProcessors
    }));
  }

  onSubmit(e) {
    if (this.props.options.onSubmit) {
      this.props.options.onSubmit();
    }
  }

  onChange(e) {
    if (this.props.options.onSubmit) {
      this.props.options.onChange();
    }
  }

  callback(e) {
    if (this.props.options.onSubmit) {
      this.props.options.callback();
    }
  }

  handleGiveConsent(e) {
    e.preventDefault();

    let data = {};
    let processors = this.state.processors.slice();

    if (processors.length < 1 && !ReactDOM.findDOMNode(this)) {
      return null;
    }

    // Get Data from parent form
    const form = ReactDOM.findDOMNode(this).parentNode;

    for (let field in { ...form }) {
      const input = form[field];
      const hasConsent = input.attributes && input.attributes.getNamedItem('data-cleargdpr');
      if (hasConsent && hasConsent.value === 'true' && input.value) {
        data[input.name] = input.value;
      }
    }

    this.giveConsent(data, processors.filter(p => p.enabled).map(p => p.id))
      .then(() => {
        e.target.submit(); // Submit parent form callback
      })
      .catch(err => {
        console.log(err);
      });
  }

  onProcessorChange(processor) {
    let processors = this.state.processors.slice();
    let index = _.findIndex(processors, { id: processor.id });
    if (index !== -1) {
      processors.splice(index, 1, processor);
      this.setState({ processors });
    }
  }

  async giveConsent(data = {}, processors) {
    // TODO: Temporary fix to store user sign up credentials
    // TODO: Should require that Elements framework store this locally with Redux
    // TODO: Should recover existent users if sign up is repeated
    if (data.email && !this.state.registered) {
      const url = API_BASE + '/api/users/register';
      const token = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          username: data.email,
          password: 'some password' //tempfix
        })
      })
        .then(res => res.json())
        .then(res => {
          if (res.error) {
            throw new Error(res.error);
          }

          return res;
        });

      localStorage.setItem('cgToken', token.cgToken);

      this.setState({
        registered: token
      });
    }

    const { cgToken } = this.state.registered;

    // TODO: Get payload from store
    const payload = {
      personalData: { ...data },
      processors: [...processors]
    };

    window.cg.setAccessToken(cgToken);
    await window.cg.Subject.giveConsent(payload)
      .then(res => {
        console.log(res);
      })
      .catch(err => {
        console.log('failure', err);
      });
  }

  render() {
    return (
      <div>
        <Checkbox
          styles={this.props.options.styles}
          label={this.props.options.label}
          onChange={this.onChange.bind(this)}
          required={this.props.options.required}
          name="cleargdpr-give-consent"
        />
        <div style={{ position: 'relative', padding: '10px 0 0 40px' }}>
          <button
            className="button is-primary is-outlined is-small"
            onClick={this.toggleProcessors.bind(this)}
          >
            <span role="img" aria-labelledby="lock">
              ðŸ”’
            </span>Config Processors
          </button>
          {this.state.showProcessors ? (
            <PopoverView
              open={this.state.showProcessors}
              onClose={this.toggleProcessors.bind(this)}
            >
              <ProcessorsList
                processors={this.state.processors}
                onProcessorChange={this.onProcessorChange.bind(this)}
              />
            </PopoverView>
          ) : null}
        </div>
      </div>
    );
  }
}

export default Consent;
