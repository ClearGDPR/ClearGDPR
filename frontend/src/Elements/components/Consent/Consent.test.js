import React from 'react';
import { mount } from 'enzyme';
import Consent from './Consent';
import PopoverView from 'components/Common/Views/Popover';
import Checkbox from 'components/Common/Checkbox';
import ProcessorsList from 'components/Processors/ProcessorsList';
import { CG } from '@cleargdpr/js-sdk';
import Subject from 'contexts/Subject';

const processors = [
  {
    id: 1,
    name: 'Nielsen',
    logoUrl: '',
    consented: 72,
    description: '',
    scopes: ['Full Name', 'Email', 'Phone Number']
  },
  {
    id: 2,
    name: 'Live Ramp',
    logoUrl: '',
    consented: 64,
    description: '',
    scopes: ['Email']
  },
  {
    id: 3,
    name: 'Experian',
    logoUrl: '',
    consented: 25,
    description: '',
    scopes: ['Full Name', 'Phone Number']
  }
];

const cg = new CG({
  apiKey: 'test',
  apiUrl: 'test'
});
cg.Subject.giveConsent = async () => {};
cg.Subject.getProcessors = async () => {
  return processors;
};

const subject = new Subject(cg, { propagateMutation: () => {} });

const defaultOptions = {
  label: `Test label`,
  styles: {
    label: {
      fontSize: '12px'
    }
  },
  required: true,
  onSuccessCallback: () => {}
};

const setup = async () => {
  return mount(<Consent options={defaultOptions} subject={subject} />);
};

const setupIntoForm = async () => {
  const Form = (
    <form>
      <input name="name" defaultValue="John" data-cleargdpr="true" />
      <input name="email" defaultValue="email@gmail.com" data-cleargdpr="true" />
      <select name="test" defaultValue="two" data-cleargdpr="true">
        <option value="one">one</option>
        <option value="two">two</option>
      </select>
      <Consent options={defaultOptions} subject={subject} />
    </form>
  );

  return mount(Form);
};

let spy;

describe('(Elements) Consent', () => {
  it('should render correctly', async () => {
    const element = await setup();
    expect(element).toMatchSnapshot();
  });

  it('should get processors list and use it for rendering', async () => {
    const element = await setup();
    expect(element.state().processors).toHaveLength(processors.length);
  });

  it('should open popover', async () => {
    const element = await setup();
    element.find('button').simulate('click');
    expect(element.find(PopoverView)).toHaveLength(1);
  });

  it('should pass some options down to checkbox component', async () => {
    const element = await setup();
    const props = element.find(Checkbox).props();
    expect(props.required).toBe(defaultOptions.required);
    expect(props.label).toBe(defaultOptions.label);
    expect(props.styles).toBe(defaultOptions.styles);
  });

  it('should listen to "onProcessorChange" event and update processors list', async () => {
    const element = await setup();
    element.find('button').simulate('click');
    const processorsList = element.find(ProcessorsList);
    expect(element.state().processors[0].enabled).toBe(true);
    processorsList.props().onProcessorChange({ ...processors[0], enabled: false });
    expect(element.state().processors[0].enabled).toBe(false);
  });

  test.skip('should give consent to all enabled processors', async () => {
    const form = await setupIntoForm();

    spy = jest.spyOn(cg.Subject, 'giveConsent');
    //make subject authenticate
    subject.isGuest = false;
    form.find(Consent).setProps({});

    expect(spy).toHaveBeenCalledWith({
      personalData: { email: 'email@gmail.com', name: 'John', test: 'two' },
      processors: [1, 2, 3]
    });
  });

  afterEach(() => {
    if (spy) {
      spy.mockClear();
    }
  });
});
