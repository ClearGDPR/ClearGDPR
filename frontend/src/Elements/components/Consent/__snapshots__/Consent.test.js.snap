// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Elements) Consent should render correctly 1`] = `
<Consent
  options={
    Object {
      "label": "Test label",
      "onSuccessCallback": [Function],
      "required": true,
      "styles": Object {
        "label": Object {
          "fontSize": "12px",
        },
      },
    }
  }
  subject={
    Subject {
      "_propagateMutation": [Function],
      "cg": v {
        "Events": y {
          "_events": Object {},
        },
        "Processors": Processors {
          "_request": s {
            "config": [Circular],
          },
        },
        "Subject": Subject {
          "_request": s {
            "config": [Circular],
          },
          "getProcessors": [Function],
          "giveConsent": [Function],
        },
        "_config": Object {
          "apiKey": "test",
          "apiTimeout": 30,
          "apiUrl": "test",
          "apiVersion": undefined,
          "auth": null,
          "debug": false,
        },
      },
      "data": null,
      "isErased": false,
      "isFetched": false,
      "isGuest": true,
      "isLoading": false,
      "objection": null,
      "processors": Array [
        Object {
          "consented": 72,
          "description": "",
          "enabled": true,
          "id": 1,
          "logoUrl": "",
          "name": "Nielsen",
          "scopes": Array [
            "Full Name",
            "Email",
            "Phone Number",
          ],
        },
        Object {
          "consented": 64,
          "description": "",
          "enabled": true,
          "id": 2,
          "logoUrl": "",
          "name": "Live Ramp",
          "scopes": Array [
            "Email",
          ],
        },
        Object {
          "consented": 25,
          "description": "",
          "enabled": true,
          "id": 3,
          "logoUrl": "",
          "name": "Experian",
          "scopes": Array [
            "Full Name",
            "Phone Number",
          ],
        },
      ],
      "restrictions": null,
      "shares": null,
      "status": null,
    }
  }
>
  <div>
    <Checkbox
      label="Test label"
      name="cleargdpr-give-consent"
      required={true}
      styles={
        Object {
          "label": Object {
            "fontSize": "12px",
          },
        }
      }
    >
      <label
        name="cleargdpr-give-consent"
      >
        <input
          required={true}
          type="checkbox"
        />
        <span
          style={
            Object {
              "fontSize": "12px",
            }
          }
        >
          Test label
        </span>
      </label>
    </Checkbox>
    <div
      style={
        Object {
          "padding": "10px 0 0 40px",
          "position": "relative",
        }
      }
    >
      <button
        className="button is-primary is-outlined is-small"
        onClick={[Function]}
      >
        <span
          aria-labelledby="lock"
          role="img"
        >
          ðŸ”’
        </span>
        Config Processors
      </button>
    </div>
  </div>
</Consent>
`;
